File: chamati/blueprints/usuarios.py
"""Gestão de usuários — somente técnico (RF01)."""
from flask import Blueprint, jsonify, request

from .. import config, db
from ..auth import login_obrigatorio

bp = Blueprint("usuarios", __name__)


@bp.route("/api/usuarios", methods=["GET"])
@login_obrigatorio(perfis=["tecnico"])
def listar():
    rows = db.consultar("SELECT email, nome, perfil FROM usuarios ORDER BY perfil, nome")
    return jsonify({"total": len(rows), "usuarios": [dict(r) for r in rows]}), 200


@bp.route("/api/usuarios", methods=["POST"])
@login_obrigatorio(perfis=["tecnico"])
def criar():
    dados = request.get_json(silent=True) or {}
    nome = (dados.get("nome") or "").strip()
    email = (dados.get("email") or "").strip().lower()
    senha = dados.get("senha") or ""
    perfil = (dados.get("perfil") or "").strip()

    erros = {}
    if len(nome) < 2:
        erros["nome"] = "Informe o nome completo."
    if "@" not in email or len(email) < 5:
        erros["email"] = "Informe um e-mail válido."
    elif db.consultar("SELECT 1 FROM usuarios WHERE email = ?", (email,), um=True):
        erros["email"] = "Já existe um usuário com este e-mail."
    if len(senha) < 6:
        erros["senha"] = "A senha deve ter ao menos 6 caracteres."
    if perfil not in config.PERFIS_VALIDOS:
        erros["perfil"] = "Selecione um perfil válido."

    if erros:
        return jsonify({"erro": "Dados invalidos", "campos": erros}), 400

    db.executar(
        "INSERT INTO usuarios (email, nome, senha, perfil) VALUES (?, ?, ?, ?)",
        (email, nome, senha, perfil),
    )
    return jsonify({
        "mensagem": "Usuário criado com sucesso",
        "usuario": {"email": email, "nome": nome, "perfil": perfil},
    }), 201

File: static/js/usuarios.js
/* Tela de gestão de usuários (somente técnico) */

const form = document.getElementById("form-usuario");
const aviso = document.getElementById("aviso");
const corpo = document.getElementById("corpo-usuarios");
const PERFIL_LABEL = { colaborador: "Colaborador", tecnico: "Técnico" };

function limparErros() {
  document.querySelectorAll(".campo.invalido").forEach((c) => c.classList.remove("invalido"));
  document.querySelectorAll(".mensagem-erro").forEach((m) => (m.textContent = ""));
  aviso.className = "aviso";
}

function marcarErro(campo, mensagem) {
  const div = document.getElementById("campo-" + campo);
  if (!div) return;
  div.classList.add("invalido");
  const msg = div.querySelector(`[data-erro="${campo}"]`);
  if (msg) msg.textContent = mensagem;
}

function validar(d) {
  const erros = {};
  if (d.nome.length < 2) erros.nome = "Informe o nome completo.";
  if (!d.email.includes("@") || d.email.length < 5) erros.email = "Informe um e-mail válido.";
  if (d.senha.length < 6) erros.senha = "A senha deve ter ao menos 6 caracteres.";
  if (!d.perfil) erros.perfil = "Selecione um perfil.";
  return erros;
}

async function carregar() {
  const { ok, dados } = await API.get("/api/usuarios");
  corpo.innerHTML = "";
  if (!ok) return;
  dados.usuarios.forEach((u) => {
    const tr = document.createElement("tr");
    tr.innerHTML = `
      <td>${escapar(u.nome)}</td>
      <td>${escapar(u.email)}</td>
      <td>${escapar(PERFIL_LABEL[u.perfil] || u.perfil)}</td>`;
    corpo.appendChild(tr);
  });
}

form.addEventListener("submit", async (e) => {
  e.preventDefault();
  limparErros();

  const dados = {
    nome: document.getElementById("nome").value.trim(),
    email: document.getElementById("email").value.trim(),
    senha: document.getElementById("senha").value,
    perfil: document.getElementById("perfil").value,
  };

  const erros = validar(dados);
  if (Object.keys(erros).length) {
    Object.entries(erros).forEach(([c, m]) => marcarErro(c, m));
    return;
  }

  const botao = document.getElementById("btn-criar");
  botao.disabled = true;
  const { ok, status, dados: resp } = await API.post("/api/usuarios", dados);
  botao.disabled = false;

  if (ok) {
    aviso.textContent = `Usuário ${resp.usuario.nome} (${PERFIL_LABEL[resp.usuario.perfil]}) criado com sucesso.`;
    aviso.className = "aviso sucesso";
    form.reset();
    carregar();
  } else if (status === 400 && resp && resp.campos) {
    Object.entries(resp.campos).forEach(([c, m]) => marcarErro(c, m));
  } else {
    aviso.textContent = (resp && resp.erro) || "Não foi possível criar o usuário.";
    aviso.className = "aviso erro";
  }
});

carregar();
